File: HappySupremacyConfig/Code/Client/HappySupremacy_RobAICFG.cpp
#include "stdafx.h"
#include "tinyxml2/tinyxml2.h"
#include "LuaCfgHelper.h"
#include <iostream>
#include <boost/algorithm/string.hpp>
#include "HappySupremacy_RobAICFG.h"
std::auto_ptr<HappySupremacy_RobAICFG> HappySupremacy_RobAICFG::msSingleton(nullptr);

int HappySupremacy_RobAICFG::GetCount()
{
	return (int)mMapData.size();
}

const HappySupremacy_RobAICFGData* HappySupremacy_RobAICFG::GetData(int ID)
{
	auto it = mMapData.find(ID);
	if (it != mMapData.end())
	{
		return &it->second;
	}
	return NULL;
}

const std::map<int, HappySupremacy_RobAICFGData>& HappySupremacy_RobAICFG::GetMapData()
{
	return mMapData;
}

void HappySupremacy_RobAICFG::Load()
{
	tinyxml2::XMLDocument xmlDoc;
	std::string content = FileUtils::getInstance()->getStringFromFile("Config/HappySupremacy_RobAICFG.xml");
	auto result = xmlDoc.Parse(content.c_str(), content.length());
	if (result != tinyxml2::XML_SUCCESS)
	{
		CCLOGERROR("Result:%d", result);
		CCASSERT(false, "result != tinyxml2::XML_SUCCESS");
		return;
	}
	auto root = xmlDoc.RootElement();
	if (root == NULL)
	{
		CCASSERT(false, "root == NULL");
		return;
	}
	auto element = root->FirstChildElement("Data");
	while (element != NULL)
	{
		HappySupremacy_RobAICFGData data;
		data.mID = element->IntAttribute("ID");
		data.mBetName = element->Attribute("BetName");
		{
			const char* BetRate = element->Attribute("BetRate");
			std::vector<std::string> vecBetRate;
			boost::split(vecBetRate, BetRate, boost::is_any_of(","));
			int temp;
			for (unsigned int i = 0; i < vecBetRate.size(); i++)
			{
				if (tinyxml2::XMLUtil::ToInt(vecBetRate[i].c_str(), &temp))
				{
					data.mBetRate.push_back(temp);
				}
			}
		}
		if (mMapData.find(data.mID) != mMapData.end())std::cout <<"data refind:" << data.mID << std::endl;
		CCASSERT(mMapData.find(data.mID) == mMapData.end(), "data.mID is exists");
		mMapData.insert(std::make_pair(data.mID, data));
		element = element->NextSiblingElement();
	}
	CCLOG("HappySupremacy_RobAICFG Loaded. Load Data:%u", mMapData.size());
}

void HappySupremacy_RobAICFG::LoadLua()
{
	LuaEngine::getInstance()->executeScriptFile("Config/HappySupremacy_RobAICFG");
	lua_State* L = LuaEngine::getInstance()->getLuaStack()->getLuaState();
	lua_getglobal(L, "HappySupremacy_RobAICFG");
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushstring(L, "datas");
	lua_gettable(L, -2);
	CCASSERT(lua_istable(L, -1) == 1, "is not table");
	lua_pushnil(L);
	while(lua_next(L, 2))
	{
		CCASSERT(lua_istable(L, -1) == 1, "is not table");
		HappySupremacy_RobAICFGData data;
		LuaCfgHelper::readInt(L, "ID", data.mID);
		LuaCfgHelper::readString(L, "BetName", data.mBetName);
		LuaCfgHelper::readVectorInt(L, "BetRate", data.mBetRate);
		if (mMapData.find(data.mID) != mMapData.end())std::cout <<"data refind:" << data.mID << std::endl;
		CCASSERT(mMapData.find(data.mID) == mMapData.end(), "data.mID is exists");
		mMapData.insert(std::make_pair(data.mID, data));
		lua_pop(L, 1);
	}
	lua_settop(L, 0);
	CCLOG("HappySupremacy_RobAICFG Loaded. Load Data:%u", mMapData.size());
}

void HappySupremacy_RobAICFG::Reload()
{
	mMapData.clear();
	Load();
}

HappySupremacy_RobAICFG* HappySupremacy_RobAICFG::GetSingleton()
{
	if (msSingleton.get() == nullptr)
	{
		msSingleton.reset(new HappySupremacy_RobAICFG());
	}
	return msSingleton.get();
}

File: HappySupremacyConfig/Lua/HappySupremacy_RobAICFG.lua
HappySupremacy_RobAICFG = {}

function HappySupremacy_RobAICFG:getData(key)
	if self.datas == nil then
		return nil
	end
	return self.datas[key]
end

function HappySupremacy_RobAICFG:init()
	self.datas = {}
	self.datas[1] = {ID = 1, BetName = "1爱好1", BetRate = {10,50,10,10,10,10}}
	self.datas[2] = {ID = 2, BetName = "1爱好2", BetRate = {15,55,5,5,10,10}}
	self.datas[3] = {ID = 3, BetName = "1爱好3", BetRate = {5,60,5,5,15,10}}
	self.datas[4] = {ID = 4, BetName = "1爱好4", BetRate = {5,65,0,0,20,10}}
	self.datas[5] = {ID = 5, BetName = "1爱好5", BetRate = {5,65,0,5,15,10}}
	self.datas[6] = {ID = 6, BetName = "2爱好1", BetRate = {10,10,10,10,50,10}}
	self.datas[7] = {ID = 7, BetName = "2爱好2", BetRate = {10,15,5,5,55,10}}
	self.datas[8] = {ID = 8, BetName = "2爱好3", BetRate = {5,15,5,5,60,10}}
	self.datas[9] = {ID = 9, BetName = "2爱好4", BetRate = {0,25,0,0,65,10}}
	self.datas[10] = {ID = 10, BetName = "2爱好5", BetRate = {0,20,5,0,65,10}}
	self.datas[11] = {ID = 11, BetName = "3爱好1", BetRate = {15,25,15,15,15,15}}
	self.datas[12] = {ID = 12, BetName = "3爱好2", BetRate = {25,15,15,15,15,15}}
	self.datas[13] = {ID = 13, BetName = "3爱好3", BetRate = {15,15,25,15,15,15}}
	self.datas[14] = {ID = 14, BetName = "3爱好4", BetRate = {15,15,15,25,15,15}}
	self.datas[15] = {ID = 15, BetName = "3爱好5", BetRate = {15,15,15,15,25,15}}
end

HappySupremacy_RobAICFG:init()
